src/fluree/db/api/transact.cljc

```diff
@@ -25,22 +30,67 @@
                                                  :fuel (fuel/tally fuel-tracker))))))))
       (<? (dbproto/-stage db json-ld opts)))))
 
+(defn- parse-json-ld-txn
+  "Expands top-level keys and parses any opts in json-ld transaction document,
+  for use by `transact!`.
+
+  Throws if required keys @id or @graph are absent."
+  [json-ld]
+  (let [context-key (cond
+                      (contains? json-ld "@context") "@context"
+                      (contains? json-ld :context) :context)
+        context (get json-ld context-key)]
+    (let [parsed-context (json-ld/parse-context context)
+          {id "@id" graph "@graph" :as parsed-txn}
+          (into {}
+                (map (fn [[k v]]
+                       (let [k* (if (= context-key k)
+                                  "@context"
+                                  (json-ld/expand-iri k parsed-context))
+                             v* (if (= const/iri-opts k*)
+                                  (keywordize-keys v)
+                                  v)]
+                         [k* v*])))
+                json-ld)]
+      (if-not (and id graph)
+        (throw (ex-info (str "Invalid transaction, missing required keys:"
+                             (when (nil? id)
+                               " @id")
+                             (when (nil? graph)
+                               " @graph")
+                             ".")
+                        {:status 400 :error :db/invalid-transaction}))
+        parsed-txn))))
+
 (defn transact!
-  [ledger json-ld opts]
+  [conn json-ld opts]
   (go-try
-    (if (:meta opts)
-      (let [start-time   #?(:clj  (System/nanoTime)
-                            :cljs (util/current-time-millis))
-            fuel-tracker (fuel/tracker)]
-        (try* (let [tx-result (<? (tx/transact! ledger fuel-tracker json-ld opts))]
-                {:status 200
-                 :result tx-result
-                 :time   (util/response-time-formatted start-time)
-                 :fuel   (fuel/tally fuel-tracker)})
-              (catch* e
-                      (throw (ex-info "Error updating ledger"
-                                      (-> e
-                                          ex-data
-                                          (assoc :time (util/response-time-formatted start-time)
-                                                 :fuel (fuel/tally fuel-tracker))))))))
-      (<? (tx/transact! ledger json-ld opts)))))
+    (let [{txn-context "@context"
+           txn "@graph"
+           ledger-id "@id"
+           txn-opts const/iri-opts
+           default-context const/iri-default-context} (parse-json-ld-txn json-ld)
+          address  (<? (conn-proto/-address conn ledger-id nil))]
+      (if-not (<? (conn-proto/-exists? conn address))
+        (throw (ex-info "Ledger does not exist" {:ledger address}))
+        (let [ledger (<? (jld-ledger/load conn address))
```

Comment:
This is probably out-of-scope for this ticket, but I think we're going to have to have some sort of state on the conn that keeps loaded ledgers around so we can look up already-loaded ledgers instead of loading on every transaction.